test_app/tests/authentication/test_social_auth.py

```diff
@@ -89,37 +90,226 @@ def test_social_auth_validate_callback_mixin(mocked_generate_slug, mocked_revers
+    'mFpbyI6IkRmMlVWWEwxaXghbE1DV01TT0pCY0ZhdHpjR2Z2Rkdoakt2OHE1ZzB4NzMyZFI1TUI1Qmlzd'
+    'kdRTzdZV0J5amQ4aVFETHEhZUdiSURha3lwNW1uT3JjZHFIZVlTbmx0ZXBRbVJwNkFJWjhqWSJ9Cg.DY'
+    'pmy4-fx4vxDlHGSpUeeyQgeyich5hijijSG8vjfIMGYXI-IobkFE7jtTp76WY-1cnzkCU_cResDY3VGk'
+    'aWzulsnbiPQJIE_tzEO1o88CZaW1rJR_KVzFli1u-03gpUwzfnzZSDTQPo13IQJ1rfAoGQMlXeQRa4gG'
+    'wREstC8fMi5DqLK_dG6BRZcYXrqRUK-4vJixJPPtS7aW70ndUsg9xbp8vW37dVU6O5hUdqC0PnKNRi5t'
+    'VJPLQF6yiserLOTI2jS_Yu6lqANdZ0pBLGi4ce5GvpEY8JbX9Rp015jQ4zOENU3_6buCGQNcUmWr9i0B'
+    'xXgzJEZpc28_SgF9I8PQ'
+)
+
+
 @pytest.mark.parametrize(
-    "groups_claim,returned_groups,expected_groups",
+    "groups_claim,returned_groups,id_token,expected_groups",
     [
-        (None, ["mygroup"], ["mygroup"]),
-        ("groups", ["mygroup"], ["mygroup"]),
-        (None, None, []),
-        ("groups", None, []),
+        (None, ["mygroup"], None, ["mygroup"]),
+        ("groups", ["mygroup"], None, ["mygroup"]),
+        (None, None, None, []),
+        ("groups", None, None, []),
+        # Check extracting groups claim from id_token
+        ("groups", None, sample_jwt, ["myidtokengroup"]),
+        # Test extracting groups claim from id_token when groups claim does not exist
+        (None, None, sample_jwt, []),
+        # Test precedence for groups claim extraction. UserInfo before id_token
+        ("groups", ["mygroup"], sample_jwt, ["mygroup"]),
+        # Test behaviour when signature verification fails
+        ("groups", None, sample_jwt_invalid_signature, []),
+        # Test behaviour when signing key can not be found
+        ("groups", None, sample_jwt_invalid_kid, []),
     ],
 )
 @mock.patch("ansible_base.authentication.utils.claims.update_user_claims")
-def test_create_user_claims_pipeline(mock_update_user_claims, groups_claim, returned_groups, expected_groups):
+def test_create_user_claims_pipeline(mock_update_user_claims, groups_claim, returned_groups, id_token, expected_groups):
     '''
     We are testing to see if extracting groups from a claim is working correctly
     '''
 
     class MockBackend(SocialAuthMixin):
         database_instance = None
 
+        JWT_ALGORITHMS = "RS256"
+        JWT_DECODE_OPTIONS = None
+
         def __init__(self, groups_claim=None):
             if groups_claim is not None:
                 self.groups_claim = groups_claim
 
+        def setting(self, a, b=None):
+            settings = {"JWT_ALGORITHMS": "RS256", "JWT_DECODE_OPTIONS": None}
+            return settings[a]
+
         def get_user_groups(self, extra_groups=[]):
             return extra_groups
 
+        def id_token_issuer(self):
+            # must match iss claim in jwt
+            return 'https://login.microsoftonline.com/9122040d-6c67-4c5b-b112-36a304b66dad/v2.0'
+
+        def get_key_and_secret(self):
+            # client_id must match aud claim in jwt
+            return ('6cb04018-a3f5-46a7-b995-940c78f5aef3', '')
+
+        def find_valid_key(self, id_token):
```

Comment:
What classes provide this method normally? Cannot find it in any superclasses of authentication plugins, though my search for it has not been exhaustive.

Reply:
Its implemented in `social_core::OpenIdConnectAuth` https://github.com/python-social-auth/social-core/blob/master/social_core/backends/open_id_connect.py#L175